README
NAME
     Daemon::Generic - framework to provide start/stop/reload for a daemon

SYNOPSIS
     use Daemon::Generic;

     sub gd_run { ... stuff }

     newdaemon();

DESCRIPTION
    Daemon::Generic provides a framework for starting, stopping,
    reconfiguring daemon-like programs. The framework provides for standard
    commands that work for as init.d files and as apachectl-like commands.

    Programs that use Daemon::Generic subclass Daemon::Generic to override
    its behavior. Almost everything that Genric::Daemon does can be
    overridden as needed.

EXAMPLE USAGE OUTPUT
    Unless overridden, the usage output for your program will look something
    like this:

     Usage: $progname [ -c file ] [ -f ] { start | stop | reload | restart | help | version | check }
      -c            Specify configuration file (defaults to $configfile)
      -f            Run in the foreground (don't detach)
      start         Starts a new $progname if there isn't one running already
      stop          Stops a running $progname
      reload        Causes a running $progname to reload it's config file.  Starts
                    a new one if none is running.
      restart       Stops a running $progname if one is running.  Starts a new one.
      check         Check the configuration file and report the daemon state
      help          Display this usage info
      version       Display the version of $progname

CONSTRUCTION
    To hand control over to "Daemon::Generic", call "newdaemon()". Control
    will be handed back through method calls to functions you define.

    Your @ISA will be modified to include "Daemon::Generic" if if it isn't
    already there.

    These are the arguments to "newdaemon()". Defaults are in (parenthesis).

    progname       ($0) The name of this program. This will be used for
                   logging and for naming the PID file.

    configfile     ("/etc/$progname.conf") The location of the configuration
                   file for this daemon.

    pidbase        (/var/run/$progname) We include the configuration file
                   name as part of the pid file in case there are multiple
                   instances of this daemon. The pidbase is is the part of
                   the PID file that does not include the configuration file
                   name.

    pidfile        ("$pidbase.$configfile.pid") The location of the process
                   id file.

    foreground     (0) Do not detach/daemon and run in the foreground
                   instead.

    debug          (0) Turn on debugging.

    no_srand       (0) Normall srand() is called. If no_srand is set then
                   srand() won't be called.

    options        () Additional arguments for Getopt::Long::GetOptions
                   which is used to parse @ARGV. Alternatively: define
                   "&gd_more_opt()".

    minimum_args   (1) Minimum number of @ARGV arguments after flags have
                   been processed.

    maximum_args   (1) Maximum number of @ARGV arguments after flags have
                   been processed.

    version        ($pkg::VERSION) The version number of the daemon.

    logpriority    Used for "logger -p".

MUST-OVERRIDE CALLBACK METHODS
    The package that subclasses Daemon::Generic must provide the following
    callback methods.

    gd_run()       This is where you put your main program.

                   It is okay to change userid/group as the first action of
                   gd_run().

MAY-OVERRIDE CALLBACK METHODS
    The package that subclasses Daemon::Generic does not have to override
    these methods but it may want to.

    gd_preconfig() "gd_preconfig()" is called to parse the configuration
                   file ("$self->{configfile}"). Preconfig is called on all
                   invocations of the daemon ("daemon reload", "daemon
                   check", "daemon stop", etc). It shouldn't start anything
                   but it can and should verify that the config file is
                   fine.

                   The return value should be a hash. With one exception,
                   the return value is only used by "gd_postconfig()". The
                   exception is that "gd_preconfig()" may return a revised
                   PID file location (key "pidfile").

                   Most uses of Daemon::Generic should define
                   "gd_preconfig".

    gd_postconfig(%config)
                   Postconfig() is called only when the daeamon is actually
                   starting up. (Or on reconfigs). It is passed the return
                   value from "gd_preconfig".

    gd_setup_signals()
                   Set things up so that SIGHUP calls gd_reconfig_event()
                   and SIGINT calls gd_quit_event(). It will call these at
                   any time so if you want to delay signal delivery or
                   something you should override this method.

    gd_getopt()    This is invoked to parse the command line. Useful things
                   to modify are:

                   $self->{configfile} The location of the configuration
                                       file to be parsed by
                                       "gd_preconfig()".

                   $self->{foreground} Run in the foreground (don't
                                       daemonize).

                   $self->{debug}      Use it yourself.

                   The supplied "gd_getopt()" method uses Getopt::Long.

    gd_parse_argv()
                   Parse any additional command line arguments beyond what
                   "gd_getopt()" handled.

                   $ARGV[0] needs to be left alone if it is one of the
                   following standard items:

                   start     Start up a new daemon.

                   stop      Stop the running daemon.

                   restart   Stop the running daemon, start a new one.

                   reload    Send a signal to the running daemon, asking it
                             to reconfigure itself.

                   check     Just check the configuration file.

                   help      Print the help screen (probably usage()).

                   version   Display the daemon's version.

                   There is no default "gd_parse_argv()".

    gd_check($pidfile, $pid)
                   Normal behavior: return. Define additional checks to run
                   when the "check" command is given. A $pid will only be
                   supplied if there is a daemon running.

    gd_version()   Normal behavior: display a version message and exit.

    gd_help()      Normal behavior: call "gd_usage()".

    gd_commands_more()
                   Used by "gd_usage()": provide information on additional
                   commands beyond "start", "stop", "reload", etc. Return is
                   an array of key value pairs.

                    sub gd_commands_more 
                    {
                           return (
                                   savestate => 'Tell xyz server to save its state',
                                   reset     => 'Tell xyz servr to reset',
                           );
                    }

    gd_flags_more  Like "gd_commands_more()" but defines additional command
                   line flags. There should also be a "gd_more_opt()" or an
                   "options" argument to "new()".

    gd_positional_more
                   Like "gd_commands_more()" but defines positional
                   arguments.

    gd_usage()     Display a usage message. The return value from
                   "gd_usage()" is the exit code for the program.

    gd_more_opt()  () Additional arguments for Getopt::Long::GetOptions
                   which is used to parse @ARGV. Alternatively: pass
                   "options" to "new()".

    gd_pidfile()   Figure out the PID file should be.

    gd_error()     Print out an error (call die?)

    gd_other_cmd() Called $ARGV[0] isn't one of the commands that
                   Daemon::Generic knows by default. Default behavior: call
                   "gd_usage()" and exit(1).

    gd_daemonize() Normal behavior: "fork()", "fork()", detach from tty.

    gd_redirect_output()
                   This is a mis-named method. Sorry. This directs
                   "STDOUT"/"STDERR"/"STDIN" to "/dev/null" as part of
                   daemonizing. Used by "gd_daemonize()".

    gd_reopen_output()
                   After daemonizing, output file descriptors are be
                   re-established. Normal behavior: redirect "STDOUT" and
                   "STDERR" to "logger -t $progname[$$]". Used by
                   "gd_daemonize()".

    gd_logname()   Normal behavior: $progname[$$]. Used by
                   "gd_redirect_output()".

    gd_reconfig_event()
                   Normal behavior: call "gd_postconfig(gd_preconfig))".
                   Only referenced by "gd_setup_signals()".

    gd_quit_event()
                   Normal behavior: exit. Only referenced by
                   "gd_setup_signals()".

    gd_kill_groups()
                   Return true if gd_kill should kill process groups ($pid)
                   instead of just the one daemon ($pid). Default is false.

    gd_kill($pid)  Used by the "stop" and "restart" commands to get rid of
                   the old daemon. Normal behavior: send a SIGINT. Check to
                   see if process $pid has died. If it has not, keep
                   checking and if it's still alive. After
                   $Daemon::Generic::force_quit_delay seconds, send a
                   SIGTERM. Keep checking. After another
                   $Daemon::Generic::force_quit_delay seconds, send a
                   SIGKILL (-9). Keep checking. After
                   "$Daemon::Generic::force_quit_delay * 4" seconds or 60
                   seconds (whichever is smaller), give up and exit(1).

    gd_install     Installs the daemon so that it runs automatically at next
                   reboot. Currently done with a symlink to $0 and
                   "/usr/sbin/update-rc.d". Please send patches for other
                   methods!

    gd_can_install Returns a function to do an "gd_install" if installation
                   is possible. Returns 0 otherwise.

    gd_install_pre($method)
                   Normal behavior: return. Called just before doing an
                   installation. The method indicates the installation
                   method (currently always "update-rc.d".)

    gd_install_post($method)
                   Normal behavior: return. Called just after doing an
                   installation.

    gd_uninstall   Will remove the daemon from the automatic startup regime.

    gd_can_uninstall
                   Returns a function to do the work for "gd_uninstall" if
                   it's possible. 0 otherwise.

    gd_uninstall_pre($method)
                   Normal behavior: return. Called just before doing an
                   un-installation. The method indicates the installation
                   method (currently always "update-rc.d".)

    gd_install_post($method)
                   Normal behavior: return. Called just after doing an
                   un-installation.

MEMBER DATA
    Since you need to subclass Daemon::Generic, you need to know what the
    internal data structures for Daemon::Generic are. With two exceptions,
    all of the member data items begin with the prefix "gd_".

    configfile     The location of the configuration file. (Not used by
                   Daemon::Generic).

    debug          Display debugging? (Not used by Daemon::Generic)

    gd_args        The original %args passed to "new".

    gd_progname    The process name. (defaults to $0)

    gd_pidfile     The location of the process ID file.

    gd_logpriority Used for "logger -p".

    gd_foreground  Are we running in the foreground?

EXAMPLE PROGRAM
     my $sleeptime = 1;

     newdaemon(
            progname        => 'ticktockd',
            pidfile         => '/var/run/ticktockd.pid',
            configfile      => '/etc/ticktockd.conf',
     );

     sub gd_preconfig
     {
            my ($self) = @_;
            open(CONFIG, "<$self->{configfile}") or die;
            while(<CONFIG>) {
                    $sleeptime = $1 if /^sleeptime\s+(\d+)/;
            }
            close(CONFIG);
            return ();
     }

     sub gd_run
     {
            while(1) {
                    sleep($sleeptime);
                    print scalar(localtime(time))."\n";
            }
     }

SEE ALSO
    With a while(1) and delayed signal delivery: Daemon::Generic::While1.

    With Event: Daemon::Generic::Event.

    With AnyEvent: Daemon::Generic::AnyEvent.

    Modules that use Daemon::Generic: SyslogScan::Daemon; IO::Event
    (rinetd.pl)

    Other modules that do similar things: Net::Daemon, Net::Server,
    Net::Server::Daemonize, NetServer::Generic, Proc::Application::Daemon,
    Proc::Daemon, Proc::Forking.

LICENSE
    Copyright (C) 2006-2010 David Muir Sharnoff <cpan@dave.sharnoff.org>.
    Copyright (C) 2011 Google, Inc. This module may be used and distributed
    on the same terms as Perl itself.

PACKAGERS
    Daemon::Generic is packaged for Fedora by Emmanuel Seyman
    <emmanuel.seyman@club-internet.fr>.



Changes
Revision history for Perl module Daemon::Generic


0.84 2013/10/04

 - Replaced CHANGELOG with Changes
 - Typo fixes contributed by David Steinbrunner

0.83 2013/04/05

 - Now using File::Flock::Forking so that locks are held on Solaris.  Bug 83085.

 - Now using $self->{gd_progname} in all messages instead of $0.  Bug 84439.

 - Added gd_kill_groups().  Bug 83294.

 - 'status' now says "No $0 running" if the pid file is missing.  Bug 83295.

 - Defining gd_preconfig is now optional.  Addresses bug 83086

 - Added gd_reopen_output() to address bug 83087.

 - Fixed bug 80604: Typos in README

0.82 2012/03/07

 - s/tmpdir/tmpdie/

0.81 2011/07/26

 - Add regression tests.

 - Add Daemon::Generic::AnyEvent

0.72 2011/05/19

 - Address https://rt.cpan.org/Public/Bug/Display.html?id=63704,
 - call setsid between forks rather than after them.

 - Address https://rt.cpan.org/Public/Bug/Display.html?id=58995,
 - don't unexpand tabs.

0.71 2010/07/14

 - Fix bug 58287 - close STDIN when daemonizing to complete the
 - detach.

0.61 2009/04/04

 - Stefan Seifert sent in a patch to make it LSB conformant.
 - Note: start now exits with 0 if the daemon is already running.

 - scips noted an error in the example program.

0.51 2007/08/15

 - Add File::Slurp dependency to Makefile.PL.   Reported
 - by Tom Lea <tom dot lea at zeninternet dot co dot uk>.

0.4 2007/03/27

 - Ronald J Kimball <rjk at tamias dot net> sent in
 - a patch to avoid accidently clobbering $0.


Makefile.PL

require 5.008;
use ExtUtils::MakeMaker;
WriteMakefile(
	NAME		=> 'Daemon::Generic',
	DISTNAME	=> 'Daemon-Generic',
	VERSION_FROM	=> 'lib/Daemon/Generic.pm',
	dist		=> { COMPRESS => 'gzip', SUFFIX => 'gz' },
	PREREQ_PM	=> {
		'Getopt::Long'		=> 0,
		'File::Flock'		=> 0,
		'File::Slurp'		=> 0,
		'Text::Wrap'		=> 0,
		'Cwd'			=> 0,
		'File::Basename'	=> 0,
	},
	BUILD_REQUIRES	=> {
		'Time::HiRes'		=> 0,
		'Event'			=> 0,
		'AnyEvent'		=> 0,
		'Eval::LineNumbers'	=> 0,
	},
	($] >= 5.005 
		?	
			(
				ABSTRACT	=> 'framework to provide start/stop/reload for a daemon',
				AUTHOR		=> 'David Muir Sharnoff <cpan@dave.sharnoff.org>',
				META_MERGE     => {
					resources  => {
						repository => 'http://github.com/muir/Daemon-Generic',
					},
				},
			) 
		: ()
	),
);

package MY;

sub postamble {
    <<'END_OF_POSTAMBLE';

pm_to_blib: README

README: lib/Daemon/Generic.pod
	pod2text  lib/Daemon/Generic.pod >README

END_OF_POSTAMBLE
}


lib/Daemon/Generic/While1.pm
package Daemon::Generic::While1;

use strict;
use warnings;
use Carp;
require Daemon::Generic;
require POSIX;
require Exporter;

our @ISA = qw(Daemon::Generic Exporter);
our @EXPORT = @Daemon::Generic::EXPORT;
our $VERSION = 0.3;

sub newdaemon
{
	local($Daemon::Generic::caller) = caller() || 'main';
	local($Daemon::Generic::package) = __PACKAGE__;
	Daemon::Generic::newdaemon(@_);
}

sub gd_setup_signals
{
	my ($self) = @_;
	$SIG{HUP} = sub {
		$self->{gd_sighup} = time;
	};
	my $child;
	$SIG{INT} = sub {
		$self->{gd_sigint} = time;
		#
		# We'll be getting a SIGTERM in a bit if we're not dead, so let's use it.
		#
		$SIG{TERM} = sub {
			$self->gd_quit_event(); 
			kill(15, $child) if $child;  # if we're still alive, let's stay that way
		};
	};
}

sub gd_sleep
{
	my ($self, $period) = @_;
	croak "Sleep period must be defined" unless defined $period;
	my $hires;
	if ($period*1000 != int($period*1000)) {
		$hires = 1;
		require Time::HiRes;
		import Time::HiRes qw(time sleep);
	}
	my $t = time;
	while (time - $t < $period) {
		return if $self->{gd_sigint};
		return if $self->{gd_sighup};
		if ($hires) {
			my $p = (time - $t < 1)
				? time - $t
				: 1;
			sleep($p);
		} else {
			sleep(1);
		}
	}
}

sub gd_run
{
	my ($self) = @_;
	while(1) {
		if ($self->{gd_sigint}) {
			$self->{gd_sigint} = 0;
			$self->gd_quit_event();
		}

		if ($self->{gd_sighup}) {
			$self->{gd_sighup} = 0;
			$self->gd_reconfig_event();
		}

		$self->gd_run_body();
	}
}

sub gd_reconfig_event
{
	my $self = shift;
	print STDERR "Reconfiguration requested\n";
	$self->gd_postconfig($self->gd_preconfig());
}

sub gd_quit_event
{
	print STDERR "Quitting...\n";
	exit(0);
}


sub gd_run_body { die "must override gd_run_body()" }

1;

=head1 NAME

 Daemon::Generic::While1 - Daemon framework with default while(1) loop

=head1 SYNOPSIS

 @ISA = qw(Daemon::Generic::While1);

 sub gd_run_body {
	# stuff
 }

=head1 DESCRIPTION

This is a slight variation on L<Daemon::Generic>: a default
C<gd_run()> provided.  It has a while(1) loop that calls 
C<gd_run_body()> over and over.  It checks for reconifg and
and terminate events and only actions them between calls
to C<gd_run_body()>. 

Terminate events will be forced through after 
C<$Daemon::Generic::force_quit_delay> seconds if
C<gd_run_body()> doesn't return quickly enough.

=head1 SUBCLASS METHODS REQUIRD

The following method is required to be overridden to subclass
Daemon::Generic::While1:

=over 15

=item gd_run_body()

This method will be called over and over.  This method should
include a call to C<sleep(1)> (or a bit more).  Reconfig events
will not interrupt it.  Quit events will only interrupt it 
after 15 seconds.  

=back

=head1 ADDITIONAL METHODS

The following additional methods are available for your use
(as compared to L<Daemon::Generic>):

=over 15

=item gd_sleep($period)

This will sleep for C<$period> seconds but in one-second
intervals so that if a SIGINT or SIGHUP arrives the sleep
period can end more quickly.

Using this makes it safe for C<gd_run_body()> to sleep for
longer than C<$Daemon::Generic::force_quit_delay> seconds 
at a time.

=back

=head1 ADDITIONAL MEMBER DATA

The following additional bits of member data are defined:

=over 15

=item gd_sigint

The time at which an (unprocessed) SIGINT was received.

=item gd_sighup

The time at which an (unprocessed) SIGHUP was received.

=back

=head1 LICENSE

Copyright (C) 2006-2010 David Muir Sharnoff <muir@idiom.com>. 
Copyright (C) 2011 Google, Inc.
This module may be used and distributed on the same terms
as Perl itself.



lib/Daemon/Generic.pm

package Daemon::Generic;

use strict;
use warnings;
require Exporter;
require POSIX;
use Getopt::Long;
use File::Slurp;
use File::Flock::Forking;
use File::Flock;
our @ISA = qw(Exporter);
our @EXPORT = qw(newdaemon);

our $VERSION = 0.83;

our $force_quit_delay = 15;
our $package = __PACKAGE__;
our $caller;

sub newdaemon
{
	my (%args) = @_;
	my $pkg = $caller || caller() || 'main';

	my $foo = bless {}, $pkg;

	unless ($foo->isa($package)) {
		no strict qw(refs);
		my $isa = \@{"${pkg}::ISA"};
		unshift(@$isa, $package);
	}

	bless $foo, 'This::Package::Does::Not::Exist';
	undef $foo;

	new($pkg, %args);
}

sub new
{
	my ($pkg, %args) = @_;

	if ($pkg eq __PACKAGE__) {
		$pkg = caller() || 'main';
	}

	srand(time ^ ($$ << 5))
		unless $args{no_srand};

	my $av0 = $0;
	$av0 =~ s!/!/.!g;

	my $self = {
		gd_av0		=> $av0,
		gd_args		=> \%args,
		gd_pidfile	=> $args{pidfile},
		gd_logpriority	=> $args{logpriority},
		gd_progname	=> $args{progname}
					? $args{progname}
					: $0,
		gd_pidbase	=> $args{pidbase}
					? $args{pidbase}
					: ($args{progname} 
						? "/var/run/$args{progname}"
						: "/var/run/$av0"),
		gd_foreground	=> $args{foreground} || 0,
		configfile	=> $args{configfile}
					? $args{configfile}
					: ($args{progname}
						? "/etc/$args{progname}.conf"
						: "/etc/$av0"),
		debug		=> $args{debug} || 0,
	};
	bless $self, $pkg;

	$self->gd_getopt;
	$self->gd_parse_argv;

	my $do = $self->{do} = $ARGV[0];

	$self->gd_help		if $do eq 'help';
	$self->gd_version	if $do eq 'version';
	$self->gd_install	if $do eq 'install';
	$self->gd_uninstall	if $do eq 'uninstall';

	$self->gd_pidfile unless $self->{gd_pidfile};

	my %newconfig = $self->gd_preconfig;

	$self->{gd_pidfile} = $newconfig{pidfile} if $newconfig{pidfile};

	print "PIDFILE=$self->{gd_pidfile}\n" if $self->{debug};

	print "Configuration looks okay\n" if $do eq 'check';

	my $pidfile = $self->{gd_pidfile};
	my $killed = 0;
	my $locked = 0;
	if (-e $pidfile) {
		if ($locked = lock($pidfile, undef, 'nonblocking')) {
			# old process is dead
			if ($do eq 'status') {
			    print "$self->{gd_progname} dead\n";
			    exit 1;
			}
		} else {
			sleep(2) if -M $pidfile < 2/86400;
			my $oldpid = read_file($pidfile);
			chomp($oldpid);
			if ($oldpid) {
				if ($do eq 'stop' or $do eq 'restart') {
					$killed = $self->gd_kill($oldpid);
					$locked = lock($pidfile);
					if ($do eq 'stop') {
						unlink($pidfile);
						exit;
					}
				} elsif ($do eq 'reload') {
					if (kill(1,$oldpid)) {
						print "Requested reconfiguration\n";
						exit;
					} else {
						print "Kill failed: $!\n";
					}
				} elsif ($do eq 'status') {
					if (kill(0,$oldpid)) {
						print "$self->{gd_progname} running - pid $oldpid\n";
						$self->gd_check($pidfile, $oldpid);
						exit 0;
					} else {
						print "$self->{gd_progname} dead\n";
						exit 1;
					}
				} elsif ($do eq 'check') {
					if (kill(0,$oldpid)) {
						print "$self->{gd_progname} running - pid $oldpid\n";
						$self->gd_check($pidfile, $oldpid);
						exit;
					} 
				} elsif ($do eq 'start' || $do eq 'debug') {
					print "\u$self->{gd_progname} is already running (pid $oldpid)\n";
					exit; # according to LSB, this is no error
				}
			} else {
				$self->gd_error("Pid file $pidfile is invalid but locked, exiting\n");
			}
		}
	} else {
		$locked = lock($pidfile, undef, 'nonblocking') 
			or die "Could not lock pid file $pidfile: $!";
	}

	if ($do eq 'reload' || $do eq 'stop' || $do eq 'check' || ($do eq 'restart' && ! $killed)) {
		print "No $self->{gd_progname} running\n";
	}

	if ($do eq 'stop') {
		unlink($pidfile);
		exit;
	}

	if ($do eq 'status') {
		print "No $self->{gd_progname} running\n";
		exit 3;
	}

	if ($do eq 'check') {
		$self->gd_check($pidfile);
		exit 
	}

	unless ($do eq 'reload' || $do eq 'restart' || $do eq 'start' || $do eq 'debug') {
		$self->gd_other_cmd($do, $locked);
	}

	unless ($self->{gd_foreground} || $do eq 'debug') {
		$self->gd_daemonize;
	}

	$locked or lock($pidfile, undef, 'nonblocking') 
		or die "Could not lock PID file $pidfile: $!";

	write_file($pidfile, "$$\n");

	print STDERR "Starting up...\n";

	$self->gd_postconfig(%newconfig);

	$self->gd_setup_signals;

	$self->gd_run;

	unlink($pidfile);
	exit(0);
}

sub gd_check {}

sub gd_more_opt { return() }

sub gd_getopt
{
	my $self = shift;
	Getopt::Long::Configure("auto_version");
	GetOptions(
		'configfile=s'	=> \$self->{configfile},
		'foreground!'	=> \$self->{gd_foreground},
		'debug!'	=> \$self->{debug},
		$self->{gd_args}{options}
			? %{$self->{gd_args}{options}}
			: (),
		$self->gd_more_opt(),
	) or exit($self->gd_usage());

	if (@ARGV < ($self->{gd_args}{minimum_args} || 1)) {
		exit($self->gd_usage());
	}
	if (@ARGV > ($self->{gd_args}{maximum_args} || 1)) {
		exit($self->gd_usage());
	}
}

sub gd_parse_argv { }

sub gd_help
{
	my $self = shift;
	exit($self->gd_usage($self->{gd_args}));
}

sub gd_version
{
	my $self = shift;
	no strict qw(refs);
	my $v = $self->{gd_args}{version} 
		|| ${ref($self)."::VERSION"} 
		|| $::VERSION 
		|| $main::VERSION 
		|| "?";
	print "$self->{gd_progname} - version $v\n";;
	exit;
} 

sub gd_pidfile
{
	my $self = shift;
	my $x = $self->{configfile};
	$x =~ s!/!.!g;
	$self->{gd_pidfile} = "$self->{gd_pidbase}$x.pid";
}

sub gd_other_cmd
{
	my $self = shift;
	$self->gd_usage;
	exit(1);
}

sub gd_redirect_output
{
	my $self = shift;
	return if $self->{gd_foreground};
	open(STDOUT, ">/dev/null") or die("open >/dev/null: $!");
	open(STDIN, "</dev/null") or die("open </dev/null: $!");
	open(STDERR, ">&STDOUT") or tmpdie("dup stdout > stderr: $!");
}

sub gd_reopen_output
{
	my $self = shift;
	return if $self->{gd_foreground};
	my $logname = $self->gd_logname;
	my $p = $self->{gd_logpriority} ? "-p $self->{gd_logpriority}" : "";
	open(STDERR, "|logger $p -t '$logname'") or tmpdie("open |logger $p -t $logname: $!");
	open(STDOUT, ">&STDERR") or tmpdie("dup stderr > stdout: $!");
	select(STDERR);
	$| = 1;
	select(STDOUT);
	$| = 1;
}

sub gd_daemonize
{
	my $self = shift;
	open(TMPERR, ">&STDERR") or die "dup STDERR > TMPERR: $!";

	print "Starting $self->{gd_progname} server\n";
	$self->gd_redirect_output();
	my $pid;
	POSIX::_exit(0) if $pid = fork;
	tmpdie("Could not fork: $!") 
		unless defined $pid;

	POSIX::setsid();

	$self->gd_reopen_output();
	print "Sucessfully daemonized\n" 
		or tmpdie("write to |logger: $!");

	close(TMPERR);
}

sub tmpdie
{
	my $msg = "@_";
	$msg .= sprintf(" at %s line %d\n", (caller())[1,2]) unless $msg =~ /\n$/;
	print TMPERR $msg;
	exit 1;
}

sub gd_logname
{
	my $self = shift;
	return $self->{gd_progname}."[$$]";
}

sub gd_reconfig_event
{
	my $self = shift;
	print STDERR "Reconfiguration requested\n";
	$self->gd_postconfig($self->gd_preconfig());
}

sub gd_quit_event
{
	my $self = shift;
	print STDERR "Quitting...\n";
	exit(0);
}

sub gd_setup_signals
{
	my $self = shift;
	$SIG{INT} = sub { $self->gd_quit_event() };
	$SIG{HUP} = sub { $self->gd_reconfig_event() };
}

sub gd_run { die "must defined gd_run()" }

sub gd_error
{
	my $self = shift;
	my $e = shift;
	my $do = $self->{do};
	if ($do && $do eq 'stop') {
		warn $e;
	} else {
		die $e;
	}
}

sub gd_flags_more { return () }

sub gd_flags
{
	my $self = shift;
	return (
		'-c file'	=> "Specify configuration file (instead of $self->{configfile})",
		'-f'		=> "Run in the foreground (don't detach)",
		$self->gd_flags_more
	);
}

sub gd_commands_more { return () }

sub gd_commands
{
	my $self = shift;
	return (
		start		=> "Starts a new $self->{gd_progname} if there isn't one running already",
		stop		=> "Stops a running $self->{gd_progname}",
		reload		=> "Causes a running $self->{gd_progname} to reload it's config file.  Starts a new one if none is running.",
		restart		=> "Stops a running $self->{gd_progname} if one is running.  Starts a new one.",
		$self->gd_commands_more(),
		($self->gd_can_install()
			? ('install' => "Setup $self->{gd_progname} to run automatically after reboot")
			: ()),
		($self->gd_can_uninstall()
			? ('uninstall' => "Do not run $self->{gd_progname} after reboots")
			: ()),
		check		=> "Check the configuration file and report the daemon state",
		help		=> "Display this usage info",
		version		=> "Display the version of $self->{gd_progname}",
		debug		=> "Starts a new $self->{gd_progname} in the foreground",
	)
}

sub gd_positional_more { return() }

sub gd_alts
{
	my $offset = shift;
	my @results;
	for (my $i = $offset; $i <= $#_; $i += 2) {
		push(@results, $_[$i]);
	}
	return @results;
}

sub gd_usage
{
	my $self = shift;

	require Text::Wrap;
	import Text::Wrap;

	my $col = 15;

	my @flags = $self->gd_flags;
	my @commands = $self->gd_commands;
	my @positional = $self->gd_positional_more;

	my $summary = "Usage: $self->{gd_progname} ";
	my $details = '';
	for my $i (gd_alts(0, @flags)) {
		$summary .= "[ $i ] ";
	}
	$summary .= "{ ";
	$summary .= join(" | ", gd_alts(0, @commands));
	$summary .= " } ";
	$summary .= join(" ", gd_alts(0, @positional));

	my (@all) = (@flags, @commands, @positional);
	while (@all) {
		my ($key, $desc) = splice(@all, 0, 2);
		local($Text::Wrap::columns) = 79;
		local($Text::Wrap::unexpand) = 0;
		$details .= wrap(
			sprintf(" %-${col}s ", $key),
			" " x ($col + 2),
			$desc);
		$details .= "\n";
	}

	print "$summary\n$details";
	return 0;
}

sub gd_install_pre {}
sub gd_install_post {}

sub gd_can_install
{
	my $self = shift;
	require File::Basename;
	my $basename = File::Basename::basename($0);
	if (
		-x "/usr/sbin/update-rc.d"
		&& 
		-x $0
		&& 
		$0 !~ m{^(?:/usr|/var)?/tmp/}
		&&
		eval { symlink("",""); 1 }
		&& 
		-d "/etc/init.d"
		&&
		! -e "/etc/init.d/$basename"
	) {
		return sub {
			$self->gd_install_pre("update-rc.d");
			require Cwd;
			my $abs_path = Cwd::abs_path($0);
			symlink($abs_path, "/etc/init.d/$basename")
				or die "Install failed: symlink /etc/init.d/$basename -> $abs_path: $!\n";
			print "+ /usr/sbin/update-rc.d $basename defaults\n";
			system("/usr/sbin/update-rc.d", $basename, "defaults");
			my $exit = $? >> 8;
			$self->gd_install_post("update-rc.d");
			exit($exit) if $exit;
		};
	}

	return 0;
}

sub gd_install
{
	my $self = shift;
	my $ifunc = $self->gd_can_install();
	die "Install command not supported\n" unless $ifunc;
	&$ifunc($self);
	exit(0);
}

sub gd_uninstall_pre {}
sub gd_uninstall_post {}

sub gd_can_uninstall
{
	my $self = shift;
	require File::Basename;
	my $basename = File::Basename::basename($0);
	require Cwd;
	my $abs_path = Cwd::abs_path($0) || 'no abs path';
	my $link = readlink("/etc/init.d/$basename") || 'no link';
	if (
		$link eq $abs_path
		&& 
		-x "/usr/sbin/update-rc.d"
	) {
		return sub {
			$self->gd_uninstall_pre("update-rc.d");
			unlink("/etc/init.d/$basename");
			print "+ /usr/sbin/update-rc.d $basename remove\n";
			system("/usr/sbin/update-rc.d", $basename, "remove");
			my $exit = $? >> 8;
			$self->gd_uninstall_post("update-rc.d");
			exit($exit) if $exit;
		}
	}
	return 0;
}

sub gd_uninstall
{
	my $self = shift;
	my $ufunc = $self->gd_can_uninstall();
	die "Cannot uninstall\n" unless $ufunc;
	&$ufunc($self);
	exit(0);
}

sub gd_kill_groups { 0 }
	

sub gd_kill
{
	my ($self, $pid) = @_;

	$pid = -abs($pid) if $self->gd_kill_groups();

	my $talkmore = 0;
	my $killed = 0;
	if (kill(0, $pid)) {
		$killed = 1;
		kill(2,$pid);
		print "Killing $pid\n";
		my $t = time;
		sleep(1) if kill(0, $pid);
		if ($force_quit_delay && kill(0, $pid)) {
			print "Waiting for $pid to die...\n";
			$talkmore = 1;
			while(kill(0, $pid) && time - $t < $force_quit_delay) {
				sleep(1);
			}
		}
		if (kill(15, $pid)) {
			print "Killing $pid with -TERM...\n";
			if ($force_quit_delay) {
				while(kill(0, $pid) && time - $t < $force_quit_delay * 2) {
					sleep(1);
				}
			} else {
				sleep(1) if kill(0, $pid);
			}
		}
		if (kill(9, $pid)) {
			print "Killing $pid with -KILL...\n";
			my $k9 = time;
			my $max = $force_quit_delay * 4;
			$max = 60 if $max < 60;
			while(kill(0, $pid)) {
				if (time - $k9 > $max) {
					print "Giving up on $pid ever dying.\n";
					exit(1);
				}
				print "Waiting for $pid to die...\n";
				sleep(1);
			}
		}
		print "Process $pid is gone\n" if $talkmore;
	} else {
		print "Process $pid no longer running\n";
	}
	return $killed;
}

sub gd_preconfig { }

sub gd_postconfig { }


1;
